query-service-impl/src/main/java/org/hypertrace/core/query/service/multivalue/MultiValueTransformation.java

```diff
@@ -0,0 +1,41 @@
+package org.hypertrace.core.query.service.multivalue;
+
+import io.reactivex.rxjava3.core.Observable;
+import io.reactivex.rxjava3.core.Single;
+import java.util.List;
+import javax.inject.Inject;
+import lombok.extern.slf4j.Slf4j;
+import org.hypertrace.core.query.service.AbstractQueryTransformation;
+import org.hypertrace.core.query.service.api.Expression;
+import org.hypertrace.core.query.service.api.Function;
+import org.slf4j.Logger;
+
+@Slf4j
+public class MultiValueTransformation extends AbstractQueryTransformation {
+  private final FunctionTransformation functionTransformation;
+
+  @Inject
+  MultiValueTransformation(FunctionTransformation functionTransformation) {
+    this.functionTransformation = functionTransformation;
+  }
+
+  @Override
+  protected Logger getLogger() {
+    return log;
+  }
+
+  @Override
+  protected Single<Expression> transformFunction(Function function) {
+    return this.transformExpressionList(function.getArgumentsList())
+        .map(function.toBuilder().clearArguments()::addAllArguments)
+        .flatMap(functionTransformation::transformFunction)
```

Comment:
This class is a pinot specific transformation, so does not belong as a query transformation, it belongs in the pinot converter. This is still at the abstract, data store agnostic level.

Reply:
Understood.
Just an implementation question here, since we are also checking the `attribute-service client` for attribute info we need a reactive way of converting in pinot converters. 
What I see right now is that since all the converters are mostly converting to PQL/SQL everything is straightforward. However if we introduce this we need a `get` on `caching client` which would return a single, so either we do a blocking get on that or we convert the entire implementation (fetching result from Pinot in the PinotRequestHandler) to a reactive impl.

Reply:
I would look at putting the logic in `PinotFunctionConverter`, which is where we go from generic function names to the pinot specific.

> if we introduce this we need a get on caching client which would return a single, so either we do a blocking get on that or we convert the entire implementation

True. I wouldn't convert the entire implementation. I'd first look to see if we can actually get this data synchronously. For example, ViewDefinition.getColumnType may give you that same information. If not, blockingGet would be acceptable here because we're already doing other blocking remote calls anyway. We don't want to do a blocking get when we're working in async code, but this is all sync code anyway.

Reply:
Explored this, seems fine except one issue. 
The `PinotFunctionConverter` and most of the request handler classes are being initialised manually (rather than guice injection) due to which we would have to change the definition of all of them (along with all the test classes) to allow the attribute client in the `PinotFunctionConverter` to be passed as a constructor argument. Is there some way to avoid this ?

Reply:
So just checked the ViewDefinition code, it is constructing the spec based on the explicit mention of whether the field is MAP/bytes (https://github.com/hypertrace/query-service/blob/1730c1ebfc6ade862f25c042a0bd2189960a99f6/query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/ViewDefinition.java#L111). 
In this case when it is multi value, one option is to implement the Attribute Client call in the view definition (https://github.com/hypertrace/query-service/blob/1730c1ebfc6ade862f25c042a0bd2189960a99f6/query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/ViewDefinition.java#L110) which again brings to our original problem of attribute client. The other option is to add an extra config in the handler for recognising multi valued columns.

Reply:
> one option is to implement the Attribute Client call in the view definition

Wouldn't do that - separate concerns (attribute vs pinot definitions).

Sounds like the option is either to thread the attribute client through, or the last thing you mentioned, add more config (would avoid this, since it's duplicate config and easy to get out of sync).  Another option could be to add that info when building the execution context - the attribute client might be available at that point and I believe we put other metadata in there.